Fix mktemp call to work inside Portacle.

Kevinpgalligan opened the pull request:
Portacle ships with BusyBox, which overrides several
common *nix commands with simpler versions, including
mktemp. This commit adds a check for Portacle and
adjusts the call to mktemp appropriately.

Tested by running `with-temp` and `cl-sat.minisat:solve`
inside Portacle, and they worked. Before, they threw an
error due to invalid call to mktemp.

It still won't work on all systems. That would require either
reviewing all versions of mktemp and adding further checks
to `with-temp`, or restricting `with-temp` to use only commands
defined by POSIX.

As per this issue: https://github.com/guicho271828/cl-sat.minisat/issues/2

Kevinpgalligan committed:
Fix mktemp call to work inside Portacle.

Portacle ships with BusyBox, which overrides several
common *nix commands with simpler versions, including
mktemp. This commit adds a check for Portacle and
adjusts the call to mktemp appropriately.

guicho271828 commented:
I believe https://stackoverflow.com/questions/2792675/how-portable-is-mktemp1 provides a rather comprehensive study.
* GNU provides `-p / --tmpdir directory`.
* Solaris supports `-p directory`
* Todd C miller's mktemp (which seems what is shipped in busybox) supports `-p  directory`
* FreeBSD (~~ OSX) has `-t prefix` but no `-p directory`
* OpenBSD has `-p directory`

So it seems `-p` is the most widely supported, except freebsd and OSX.

Two approach: (assuming the raw freebsd users are small,) it would be viable to ask OSX users to install coreutils (from brew) and we use `gmktemp` on OSX based on features.

Another is to implement the function with `-t prefix` on freebsd/osx.

Kevinpgalligan commented:
I'll take another look when I get the chance. I've been distracted because now that it works in Portacle, I'm unblocked from working on my own project.

Kevinpgalligan commented:
Hi, I'm finally free to work on this. Just a question: would you be opposed to removing the direct system calls and using CL library functions instead? All we would need is a function to create a directory, a function to create a file, and a function to generate a random string. I suspect that it might be more portable and future-proof than trying to cover all of the edge cases with specific CLI commands.

guicho271828 commented:
If there is a function like that then I will appreciate it

guicho271828 commented:
> generate a random string
well yes you can reimplement mktemp, which guarantees making a *new* temporary file with a random suffix. Such program must "check for existence, create if not exist, retry with a random name  if exist" in an atomic way (i.e. guarantee that another program creates the file after the check before the creation). `mktemp` uses `open(2)` unix system call with `O_EXCL` which guarantees this behavior. I believe `:if-exists :error` of `cl:open` has the same functionality (and is actually implemented as `open(2)` + `O_EXCL`) and therefore also atomic. There might be a similar function in uiop etc.

Kevinpgalligan committed:
Make with-temp more portable.

Using native Common Lisp to do it instead of
calling mktemp through uiop:run-program, which had
portability issues.

Had to implement it from scratch, couldn't find an
equivalent in uiop / other libraries that works with
files AND directories.

Kevinpgalligan commented:
I couldn't find an equivalent of what we wanted in uiop or other libraries, so I had to implement it from scratch. I tried to change the interface of `with-temp` as little as possible. Let me know if there are any changes we should make.

Here's the testing I did, starting from the functions at the bottom of the call chain.

```lisp
SAT> (create-nonexisting nil "/tmp/somefile")
T
SAT> (create-nonexisting nil "/tmp/somefile")
NIL
SAT> (create-nonexisting t "/tmp/somedir/")
T
SAT> (create-nonexisting t "/tmp/somedir/")
NIL
SAT> (random-base62 5)
"WtIVw"
SAT> (generate-temp-name "prefix." 5)
"prefix.fhI0C"
SAT> (attempt-create-temp nil "/tmp/" "prefix." 5 3)
"/tmp/prefix.hE8CI"
SAT> (attempt-create-temp nil "/tmp/" "gonnafailonsecondtry" 0 3)
"/tmp/gonnafailonsecondtry"
SAT> (attempt-create-temp nil "/tmp/" "gonnafailonsecondtry" 0 3)
; Evaluation aborted on #<SIMPLE-ERROR "Couldn't create a unique temp file/folder." {1002BA2A33}>.
SAT> (attempt-create-temp t "/tmp/" "a-dir" 5 3)
"/tmp/a-dirXk7td/"
SAT> (with-temp (f)
       f)
"/tmp/tmp.zPcyCos"
SAT> (with-temp (f :template "hello.XXXXX")
       f)
"/tmp/hello.CnGoJ"
SAT> (with-temp (f :debug t)
       f)
not removing /tmp/tmp.w7E4gdA for debugging
"/tmp/tmp.w7E4gdA"
SAT> (uiop:run-program "ls /tmp/ | grep tmp.w" :output '(:string :stripped t))
"tmp.w7E4gdA"
NIL
0
SAT> (uiop:run-program "ls /tmp/ | grep hello." :output '(:string :stripped t) :ignore-error-status t)
""
NIL
1
SAT> (with-temp (f :debug t :directory t)
       f)
not removing /tmp/tmp.9tsLIx0/ for debugging
"/tmp/tmp.9tsLIx0/"
SAT> (uiop:run-program "ls /tmp/ | grep tmp.9ts." :output '(:string :stripped t) :ignore-error-status t)
"tmp.9tsLIx0"
NIL
0
SAT> 
```

guicho271828 commented on src/1-util.lisp:
```diff
@@ -1,32 +1,68 @@
+          (progn ,@body)
        (if ,debug
            (format t "~&not removing ~a for debugging" ,var)
            (uiop:run-program (format nil "rm -rf ~a" (namestring ,var)) :ignore-error-status t)))))
```
now that we are lispifying it, can we make it just use `uiop:delete-directory-tree` ?

guicho271828 commented on src/1-util.lisp:
```diff
@@ -1,32 +1,68 @@
+  "Creates a file/directory in BASE-DIR with NAME-PREFIX as a prefix of the name and RANDOM-STRING-SIZE
+   random base62 characters at the end.
+   Returns the name of the created file.
+   Signals an error if it can't generate a unique name after ATTEMPTS attempts."
```
As per https://github.com/coreutils/coreutils/blob/master/src/mktemp.c#L260 , can we check the minimum value for `random-string-size`?

Kevinpgalligan replied:
👍🏻

guicho271828 commented on src/1-util.lisp:
```diff
@@ -1,32 +1,68 @@
+                                   tmpdir
+                                   template-without-xs
+                                   (- (length template) (length template-without-xs))
+                                   3))))
```
I think you can make this limit larger. Coreutils use 62^3 and therefore requires minimum three Xs in the template. https://fossies.org/dox/coreutils-8.31/tempname_8c_source.html

```
   /* A lower bound on the number of temporary files to attempt to
      generate.  The maximum total number of temporary file names that
      can exist for a given template is 62**6.  It should never be
      necessary to try all of these combinations.  Instead if a reasonable
      number of names is tried (we define reasonable as 62**3) fail to
      give the system administrator the chance to remove the problems.
      This value requires that X_SUFFIX_LEN be at least 3.  */
 #define ATTEMPTS_MIN (62 * 62 * 62)
```

Kevinpgalligan replied:
Makes sense.

Side note, what's the best way to define constants in CL? I tried it in my other project but it always resulted in warnings / errors about redefining symbols. Even with `alexandria:define-constant`.

guicho271828 replied:
you could put them in a separate file and load the file first before use.

Kevinpgalligan commented:
Thanks for the feedback, I should have time to work on it tomorrow.

Kevinpgalligan committed:
Fix comments from review.

1. Remove "rm -rf", replace with something Lispy.
2. Mimic behaviour of tempname from coreutils, throwing
   an error if the random component of the temp name isn't
   long enough and making more attempts in case of name
   collisions.

Kevinpgalligan commented:
Hi, I've addressed your comments. Here's the extra testing I've done.

```lisp
;; Enforcing minimum length of random component.
SAT> (with-temp (f :template "tmp.XX") f)
; Evaluation aborted on #<SIMPLE-ERROR "Random string part of temporary file name isn't long enough." {10042B3EC3}>.
;; Lispy file/directory deletion.
SAT> (uiop:probe-file* (with-temp (f) f))
NIL
SAT> (uiop:probe-file* (with-temp (f :debug t) f))
not removing /tmp/tmp.12gDbtx for debugging
#P"/tmp/tmp.12gDbtx"
SAT> (uiop:probe-file* (with-temp (f :directory t) f))
NIL
SAT> (uiop:probe-file* (with-temp (f :directory t :debug t) f))
not removing /tmp/tmp.UVpE0gM/ for debugging
#P"/tmp/tmp.UVpE0gM/"
```

One concern: I'm getting weird compile notes that I don't understand, not sure if it's an issue.

```
2 compiler notes:

2-class.lisp:23:3:
  note: 
    deleting unreachable code
    --> LET UNWIND-PROTECT FLET #:CLEANUP-FUN-9 BLOCK IF IF 
    --> UIOP/FILESYSTEM:DELETE-DIRECTORY-TREE MAKE-PATHNAME LIST 
    ==>
      CL-SAT::TMP
    

4-competition.lisp:70:3:
  note: 
    deleting unreachable code
    --> LET UNWIND-PROTECT FLET #:CLEANUP-FUN-62 BLOCK IF IF DELETE-FILE 
    ==>
      CL-SAT::DIR
```

Kevinpgalligan commented:
Hey! I hope you're keeping well, what with the COVID-19 stuff happening.

I'm sorry for dragging this out for so long, I've been really slow. Any chance of another review, though? I'm pretty sure the code is working okay, I just can't figure out how to get rid of those compiler warnings. If you have any suggestions, it would be appreciated.

guicho271828 commented:
ah no problem. I didn't notice the new patch, thank you!

guicho271828 commented:
I also think it is all set. Thank you for the work!

Kevinpgalligan commented:
Thanks for making this thing in the first place!
